File: backend/shared/dto/simulations/SimulationResponse.dto.ts
// 분위수 결과 정보 인터페이스 - 모든 수치는 원 단위 정수(Amount)로 반환
export interface PercentileResultDto {
  p10: number; // 하위 10%
  p50: number; // 중위 50% (기대값)
  p90: number; // 상위 10%
}

// 목표 달성 분석 결과 정보 인터페이스
export interface GoalAchievementResultDto {
  // 목표 도달에 소요된 예상 개월 수 - 분석 한계치(600개월) 내에 도달하지 못한 경우 null을 반환
  expectedMonthsToTarget: number | null;
}

export class SimulationResponseDto {
  public percentiles: {
    portfolioValue: PercentileResultDto;
    monthlyDividend: PercentileResultDto;
  };

  public goalAnalysis: {
    portfolioValueGoal: GoalAchievementResultDto;
    monthlyDividendGoal: GoalAchievementResultDto;
  };

  constructor(params: {
    percentiles: {
      portfolioValue: PercentileResultDto;
      monthlyDividend: PercentileResultDto;
    };
    goalAnalysis: {
      portfolioValueGoal: GoalAchievementResultDto;
      monthlyDividendGoal: GoalAchievementResultDto;
    };
  }) {
    this.percentiles = params.percentiles;
    this.goalAnalysis = params.goalAnalysis;
  }

  // 서비스 레이어의 결과 데이터를 DTO 구조로 변환하는 정적 메서드
  public static fromResults(
    mcResults: {
      portfolioValue: PercentileResultDto;
      monthlyDividend: PercentileResultDto;
    },
    goalResults: {
      portfolioValueReachedMonths: number | null;
      monthlyDividendReachedMonths: number | null;
    }
  ): SimulationResponseDto {
    return new SimulationResponseDto({
      percentiles: {
        portfolioValue: mcResults.portfolioValue,
        monthlyDividend: mcResults.monthlyDividend,
      },
      goalAnalysis: {
        portfolioValueGoal: {
          expectedMonthsToTarget: goalResults.portfolioValueReachedMonths,
        },
        monthlyDividendGoal: {
          expectedMonthsToTarget: goalResults.monthlyDividendReachedMonths,
        },
      },
    });
  }
}

File: apps/mobile/lib/features/simulation/types/simulation_response_dto.dart
class SimulationResponseDto {
  final SimulationPercentilesDto percentiles;
  final SimulationGoalAnalysisDto goalAnalysis;

  SimulationResponseDto({
    required this.percentiles,
    required this.goalAnalysis,
  });

  factory SimulationResponseDto.fromJson(Map<String, dynamic> json) {
    return SimulationResponseDto(
      percentiles: SimulationPercentilesDto.fromJson(
        json['percentiles'] as Map<String, dynamic>,
      ),
      goalAnalysis: SimulationGoalAnalysisDto.fromJson(
        json['goalAnalysis'] as Map<String, dynamic>,
      ),
    );
  }
}

class SimulationPercentilesDto {
  final SimulationPercentileValueDto portfolioValue;
  final SimulationPercentileValueDto monthlyDividend;

  SimulationPercentilesDto({
    required this.portfolioValue,
    required this.monthlyDividend,
  });

  factory SimulationPercentilesDto.fromJson(Map<String, dynamic> json) {
    return SimulationPercentilesDto(
      portfolioValue: SimulationPercentileValueDto.fromJson(
        json['portfolioValue'] as Map<String, dynamic>,
      ),
      monthlyDividend: SimulationPercentileValueDto.fromJson(
        json['monthlyDividend'] as Map<String, dynamic>,
      ),
    );
  }
}

class SimulationPercentileValueDto {
  final double p10;
  final double p50;
  final double p90;

  SimulationPercentileValueDto({
    required this.p10,
    required this.p50,
    required this.p90,
  });

  factory SimulationPercentileValueDto.fromJson(Map<String, dynamic> json) {
    return SimulationPercentileValueDto(
      p10: (json['p10'] as num).toDouble(),
      p50: (json['p50'] as num).toDouble(),
      p90: (json['p90'] as num).toDouble(),
    );
  }
}

class SimulationGoalAnalysisDto {
  final SimulationGoalMetricDto portfolioValueGoal;
  final SimulationGoalMetricDto monthlyDividendGoal;

  SimulationGoalAnalysisDto({
    required this.portfolioValueGoal,
    required this.monthlyDividendGoal,
  });

  factory SimulationGoalAnalysisDto.fromJson(Map<String, dynamic> json) {
    return SimulationGoalAnalysisDto(
      portfolioValueGoal: SimulationGoalMetricDto.fromJson(
        json['portfolioValueGoal'] as Map<String, dynamic>,
      ),
      monthlyDividendGoal: SimulationGoalMetricDto.fromJson(
        json['monthlyDividendGoal'] as Map<String, dynamic>,
      ),
    );
  }
}

class SimulationGoalMetricDto {
  final int? expectedMonthsToTarget;

  SimulationGoalMetricDto({required this.expectedMonthsToTarget});

  factory SimulationGoalMetricDto.fromJson(Map<String, dynamic> json) {
    final value = json['expectedMonthsToTarget'];

    return SimulationGoalMetricDto(
      expectedMonthsToTarget: value == null ? null : (value as num).toInt(),
    );
  }
}
